nibabel/filebasedimages.py

```diff
@@ -0,0 +1,569 @@
+    holds active file-like objects, then these may be affected by the
+    initial file read; in this case, the contains file-like objects need to
+    carry the position at which a write (with ``to_files``) should place the
+    data.  The ``file_map`` contents should therefore be such, that this will
+    work:
+
+       >>> # write an image to files
+       >>> from io import BytesIO
+       >>> file_map = nib.AnalyzeImage.make_file_map()
+       >>> file_map['image'].fileobj = BytesIO()
+       >>> file_map['header'].fileobj = BytesIO()
+       >>> img = nib.AnalyzeImage(data, np.eye(4))
+       >>> img.file_map = file_map
+       >>> img.to_file_map()
+       >>> # read it back again from the written files
+       >>> img2 = nib.AnalyzeImage.from_file_map(file_map)
+       >>> np.all(img2.get_data() == data)
+       True
+       >>> # write, read it again
+       >>> img2.to_file_map()
+       >>> img3 = nib.AnalyzeImage.from_file_map(file_map)
+       >>> np.all(img3.get_data() == data)
+       True
+    '''
+    files_types = (('image', None),)
+    alternate_exts = ()  # Modified by @ImageOpener.register_ext_from_image
+    _compressed_suffixes = ()
+    header_class = FileBasedHeader
+
+    makeable = True  # Used in test code
+    rw = True  # Used in test code
+
+    def __init__(self, header=None, extra=None, file_map=None):
+        ''' Initialize image
+
+        The image is a combination of (header), with
+        optional metadata in `extra`, and filename / file-like objects
+        contained in the `file_map` mapping.
+
+        Parameters
+        ----------
+        header : None or mapping or header instance, optional
+           metadata for this image format
+        extra : None or mapping, optional
+           metadata to associate with image that cannot be stored in the
+           metadata of this image type
+        file_map : mapping, optional
+           mapping giving file information for this image format
+        '''
+
+        if header or self.header_class:
+            self._header = self.header_class.from_header(header)
+        else:
+            self._header = None
+        if extra is None:
+            extra = {}
+        self.extra = extra
+
+        if file_map is None:
+            file_map = self.__class__.make_file_map()
+        self.file_map = file_map
+
+    @property
+    def header(self):
+        return self._header
+
+    def __getitem__(self):
+        ''' No slicing or dictionary interface for images
+        '''
+        raise TypeError("Cannot slice image objects; consider slicing image "
+                        "array data with `img.dataobj[slice]` or "
+                        "`img.get_data()[slice]`")
+
+    def get_header(self):
+        """ Get header from image
+
+        Please use the `header` property instead of `get_header`; we will
+        deprecate this method in future versions of nibabel.
+        """
+        return self.header
+
+    def get_filename(self):
```

Comment:
This one I want @matthew-brett's comment on. Since other get/sets in the vicinity seem to have been deprecated, there might be a good reason this one hasn't been.